src/Language/PureScript/Ide/Types.hs

```diff
@@ -31,17 +33,64 @@ import           Text.Parsec.Text
 type ModuleIdent = Text
 
 data IdeDeclaration
-  = IdeValue P.Ident P.Type
-  | IdeType (P.ProperName 'P.TypeName) P.Kind
-  | IdeTypeSynonym (P.ProperName 'P.TypeName) P.Type
-  | IdeDataConstructor (P.ProperName 'P.ConstructorName) (P.ProperName 'P.TypeName) P.Type
-  | IdeTypeClass (P.ProperName 'P.ClassName)
-  | IdeValueOperator (P.OpName 'P.ValueOpName) (P.Qualified (Either P.Ident (P.ProperName 'P.ConstructorName))) P.Precedence P.Associativity (Maybe P.Type)
-  | IdeTypeOperator (P.OpName 'P.TypeOpName) (P.Qualified (P.ProperName 'P.TypeName)) P.Precedence P.Associativity (Maybe P.Kind)
+  = IdeDeclValue IdeValue
+  | IdeDeclType IdeType
+  | IdeDeclTypeSynonym IdeSynonym
+  | IdeDeclDataConstructor IdeDataConstructor
+  | IdeDeclTypeClass (P.ProperName 'P.ClassName)
+  | IdeDeclValueOperator IdeValueOperator
+  | IdeDeclTypeOperator IdeTypeOperator
   deriving (Show, Eq, Ord)
 
-data IdeDeclarationAnn = IdeDeclarationAnn Annotation IdeDeclaration
-  deriving (Show, Eq, Ord)
+data IdeValue = IdeValue
+  { _ideValueIdent :: P.Ident
+  , _ideValueType :: P.Type
+  } deriving (Show, Eq, Ord)
+
+data IdeType = IdeType
+ { _ideTypeName :: P.ProperName 'P.TypeName
+ , _ideTypeKind :: P.Kind
+ } deriving (Show, Eq, Ord)
+
+data IdeSynonym = IdeSynonym
+  { _ideSynonymName :: P.ProperName 'P.TypeName
+  , _ideSynonymType :: P.Type
+  } deriving (Show, Eq, Ord)
+
+data IdeDataConstructor = IdeDataConstructor
+  { _ideDtorName :: P.ProperName 'P.ConstructorName
+  , _ideDtorTypeName :: P.ProperName 'P.TypeName
+  , _ideDtorType :: P.Type
+  } deriving (Show, Eq, Ord)
+
+data IdeValueOperator = IdeValueOperator
+  { _ideValueOpName :: P.OpName 'P.ValueOpName
+  , _ideValueOpAlias :: P.Qualified (Either P.Ident (P.ProperName 'P.ConstructorName))
+  , _ideValueOpPrecedence :: P.Precedence
+  , _ideValueOpAssociativity :: P.Associativity
+  , _ideValueOpType :: Maybe P.Type
+  } deriving (Show, Eq, Ord)
+
+data IdeTypeOperator = IdeTypeOperator
+  { _ideTypeOpName :: P.OpName 'P.TypeOpName
+  , _ideTypeOpAlias :: P.Qualified (P.ProperName 'P.TypeName)
+  , _ideTypeOpPrecedence :: P.Precedence
+  , _ideTypeOpAssociativity :: P.Associativity
+  , _ideTypeOpKind :: Maybe P.Kind
+  } deriving (Show, Eq, Ord)
+
+makePrisms ''IdeDeclaration
+makeLenses ''IdeValue
+makeLenses ''IdeType
+makeLenses ''IdeSynonym
+makeLenses ''IdeDataConstructor
+makeLenses ''IdeValueOperator
+makeLenses ''IdeTypeOperator
+
+data IdeDeclarationAnn = IdeDeclarationAnn
+  { _idaAnnotation :: Annotation
+  , _idaDeclaration :: IdeDeclaration
```

Comment:
Oh yeah now I see how that looks confusing^^ I'd love to drop all these prefixes and use the new OverloadedRecord thingy but I've heard it produces horrible error messages and we're not all GHC 8 yet.